frontend/src/pages/NewRun.test.tsx

```diff
@@ -0,0 +1,1160 @@
+  });
+
+  it('does not include any action buttons in the toolbar', async () => {
+    const props = generateProps();
+    // Clear the experiment ID from the query params, as it used at some point to update the
+    // breadcrumb, and we cover that in a later test.
+    props.location.search = '';
+
+    const tree = shallow(<TestNewRun {...props as any} />);
+    await TestUtils.flushPromises();
+
+    expect(updateToolbarSpy).toHaveBeenLastCalledWith({
+      actions: [],
+      breadcrumbs: [
+        { displayName: 'Experiments', href: RoutePage.EXPERIMENTS },
+        { displayName: 'Start a new run', href: '' }
+      ],
+    });
+    tree.unmount();
+  });
+
+  it('clears the banner when refresh is called', async () => {
+    const tree = shallow(<TestNewRun {...generateProps() as any} />);
+    expect(updateBannerSpy).toHaveBeenCalledTimes(1);
+    (tree.instance() as TestNewRun).refresh();
+    await TestUtils.flushPromises();
+    expect(updateBannerSpy).toHaveBeenCalledTimes(2);
+    expect(updateBannerSpy).toHaveBeenLastCalledWith({});
+  });
+
+  it('clears the banner when load is called', async () => {
+    const tree = shallow(<TestNewRun {...generateProps() as any} />);
+    expect(updateBannerSpy).toHaveBeenCalledTimes(1);
+    (tree.instance() as TestNewRun).load();
+    await TestUtils.flushPromises();
+    expect(updateBannerSpy).toHaveBeenCalledTimes(2);
+    expect(updateBannerSpy).toHaveBeenLastCalledWith({});
+  });
+
+  it('allows updating the run name', async () => {
+    const tree = shallow(<TestNewRun {...generateProps() as any} />);
+    await TestUtils.flushPromises();
+
+    (tree.instance() as any).handleChange('runName')({ target: { value: 'run name' } });
+
+    expect(tree.state()).toHaveProperty('runName', 'run name');
+    tree.unmount();
+  });
+
+  it('allows updating the run description', async () => {
+    const tree = shallow(<TestNewRun {...generateProps() as any} />);
+    await TestUtils.flushPromises();
+    (tree.instance() as any).handleChange('description')({ target: { value: 'run description' } });
+
+    expect(tree.state()).toHaveProperty('description', 'run description');
+    tree.unmount();
+  });
+
+  it('exits to the AllRuns page if there is no associated experiment', async () => {
+    const props = generateProps();
+    // Clear query params which might otherwise include an experiment ID.
+    props.location.search = '';
+
+    const tree = shallow(<TestNewRun {...props as any} />);
+    await TestUtils.flushPromises();
+    tree.find('#exitNewRunPageBtn').simulate('click');
+
+    expect(historyPushSpy).toHaveBeenCalledWith(RoutePage.RUNS);
+    tree.unmount();
+  });
+
+  it('fetches the associated experiment if one is present in the query params', async () => {
+    const experiment = newMockExperiment();
+    const props = generateProps();
+    props.location.search = `?${QUERY_PARAMS.experimentId}=${experiment.id}`;
+
+    getExperimentSpy.mockImplementation(() => experiment);
```

Comment:
sgtm. I think verbosity is good until it can cause problems if things are forgotten. We make certain assumptions about resetting state for each test, including clearing mocks and returning mock objects, so if we're not doing this for everything about the test state, we can easily assume something is being automatically set when it's not, and forget to do it in the test.